core/commands/storage/upload/upload/renew_init.go

```diff
@@ -0,0 +1,196 @@
+		if err != nil {
+			return err
+		}
+		guardContractMeta := halfSignedGuardContract.ContractMeta
+		pid, ok := remote.GetStreamRequestRemotePeerID(req, ctxParams.N)
+		if !ok {
+			return fmt.Errorf("fail to get peer ID from request")
+		}
+		var peerId string
+		if peerId = pid.String(); len(req.Arguments) >= 7 {
+			peerId = req.Arguments[6]
+		}
+		payerPubKey, err := crypto.GetPubKeyFromPeerId(peerId)
+		if err != nil {
+			return err
+		}
+		s := halfSignedGuardContract.GetRenterSignature()
+		if s == nil {
+			s = halfSignedGuardContract.GetPreparerSignature()
+		}
+		ok, err = crypto.Verify(payerPubKey, &guardContractMeta, s)
+		if !ok || err != nil {
+			return fmt.Errorf("can't verify guard contract: %v", err)
+		}
+		signedGuardContract, err := signGuardContract(&guardContractMeta, halfSignedGuardContract, ctxParams.N.PrivateKey)
+		if err != nil {
+			return err
+		}
+		signedGuardContractBytes, err := proto.Marshal(signedGuardContract)
+		if err != nil {
+			return err
+		}
+		go func() {
+			tmp := func() error {
+				shard, err := sessions.GetHostShard(ctxParams, signedGuardContract.ContractId)
+				if err != nil {
+					return err
+				}
+				_, err = remote.P2PCall(ctxParams.Ctx, ctxParams.N, requestPid, "/storage/upload/recvcontract",
+					ssId,
+					shardHash,
+					shardIndex,
+					nil,
+					signedGuardContractBytes,
+				)
+				if err != nil {
+					return err
+				}
+
+				signedContractID, err := signContractID(signedGuardContract.ContractId, ctxParams.N.PrivateKey)
+				if err != nil {
+					return err
+				}
+
+				paidIn := make(chan bool)
+				go checkPaymentFromClient(ctxParams, paidIn, signedContractID)
+				paid := <-paidIn
+				if !paid {
+					return fmt.Errorf("contract is not paid: %s", signedGuardContract.ContractId)
+				}
+
+				err = shard.Contract(nil, signedGuardContract)
+				if err != nil {
+					return err
+				}
+				in := &guardpb.ReadyForChallengeRequest{
+					RenterPid:   guardContractMeta.RenterPid,
+					FileHash:    guardContractMeta.FileHash,
+					ShardHash:   guardContractMeta.ShardHash,
+					ContractId:  guardContractMeta.ContractId,
+					HostPid:     guardContractMeta.HostPid,
+					PrepareTime: guardContractMeta.RentStart,
+				}
+				sign, err := crypto.Sign(ctxParams.N.PrivateKey, in)
+				if err != nil {
+					return err
+				}
+				in.Signature = sign
+				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+				defer cancel()
+				err = grpc.GuardClient(ctxParams.Cfg.Services.GuardDomain).WithContext(ctx,
+					func(ctx context.Context, client guardpb.GuardServiceClient) error {
+						_, err = client.ReadyForChallenge(ctx, in)
+						if err != nil {
+							return err
+						}
+						return nil
+					})
+				if err != nil {
+					return err
+				}
+				if err := shard.Complete(); err != nil {
+					return err
+				}
+				return nil
+			}()
+			if tmp != nil {
+				log.Error(tmp)
+			}
+		}()
+		return cmds.EmitOnce(res, signedGuardContract)
```

Comment:
mostly duplicate from core/commands/storage/upload/upload/init.go, should combine them, and use a flag to control renew or not.

Reply:
ok, let me think the reject contract implementation before merging.

Reply:
ok, sure